teps/0010-optional-workspaces.md

```diff
@@ -0,0 +1,354 @@
+Binding, even if one is configured in the `config-defaults` ConfigMap.
+
+Optional Workspaces that receive a binding from a TaskRun are validated in the
+same ways that required Workspaces are.
+
+A variable is provided that resolves to "true" or "false", indicating whether
+an optional Workspace was bound or not:
+
+```yaml
+# task.yaml
+spec:
+  workspaces:
+  - name: test-results
+    optional: true
+  script: |
+    if [ $(workspaces.test-results.bound) == "true" ] ; then
+      # write verbose test result output to workspace
+    fi
+```
+
+Required Workspaces also provide the variable indicating bound state but the
+value is always `"true"`.
+
+When an optional Workspace is omitted by a TaskRun there should be no
+placeholder mounted in its absence. This does not preclude mechanisms like
+`creds-init` from populating that part of the filesystem though. The
+rationale for this is that checks in the Task like `if [ -d $HOME/.ssh ]`
+should accurately reflect whether content was placed there or not, be it
+from a bound Workspace, `creds-init` etc. An omitted Workspace must not
+result in an empty directory being placed in its absence, resulting in an
+incorrectly passing behaviour gate.
+
+### Example: `git-clone` Catalog Task
+
+The `git-clone` Catalog Task already supports the `creds-init` approach to
+credential handling. Here is an example updating the Task to include support
+for an explicit `.ssh` directory to be provided as an optional Workspace:
+
+```yaml
+# A number of fields which have no bearing on the behaviour of this example
+# have been removed from the original Catalog entry.
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: git-clone
+spec:
+  workspaces:
+    - name: output
+      description: The git repo will be cloned onto the volume backing this workspace
+    - name: ssh-directory
+      description: An .ssh directory with private key and known_hosts file.
+      optional: true
+  params:
+    - name: url
+      description: git url to clone
+      type: string
+    - name: revision
+      description: git revision to checkout (branch, tag, sha, ref…)
+      type: string
+      default: master
+    - name: subdirectory
+      description: subdirectory inside the "output" workspace to clone the git repo into
+      type: string
+      default: ""
+  results:
+    - name: commit
+      description: The precise commit SHA that was fetched by this Task
+  steps:
+    - name: clone
+      image: gcr.io/tekton-releases/github.com/tektoncd/pipeline/cmd/git-init:v0.14.2
+      script: |
+        CHECKOUT_DIR="$(workspaces.output.path)/$(params.subdirectory)"
+
+        # Here we only use the credentials provided via the ssh-directory Workspace
+        # if creds-init did not initialize any SSH creds already.
+        if [ ! -d "$HOME/.ssh" ] ; then
+          if [ "$(workspaces.ssh-directory.bound)" == "true" ]; then
+            cp -R "$(workspaces.ssh-directory.path)" "$HOME/.ssh"
```

Comment:
what would be the value for `$(workspaces.ssh-directory.path)` in the case the workspace is not bound ? I guess empty but I think we should document this.

Reply:
Ah good catch.  I included the following but forgot to also mention the solution to it:

```
Tekton Variables should fail gracefully in the event that a workspace is
not provided. Using a variable like `$(workspaces.foo.path)` should not
cause the controller to abandon or fail a TaskRun when the optional workspace
`foo` is omitted by it. Likewise the literal string `"$(workspaces.foo.path)"`
should not end up being passed to the Task.
```

You're right - I'm thinking of injecting an empty string for the path if an optional workspace is not provided.

Reply:
I've updated the Proposal to describe the behaviour of variables.